hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileReaderImpl.java

```diff
@@ -1357,9 +1357,10 @@ public HFileBlock readBlock(long dataBlockOffset, long onDiskBlockSize, final bo
         BlockType.BlockCategory category = hfileBlock.getBlockType().getCategory();
         final boolean cacheCompressed = cacheConf.shouldCacheCompressed(category);
         final boolean cacheOnRead = cacheConf.shouldCacheBlockOnRead(category);
+        final boolean shouldCacheFileBlock = cacheConf.shouldCacheFileBlock(getHFileInfo(), conf);
 
         // Don't need the unpacked block back and we're storing the block in the cache compressed
-        if (cacheOnly && cacheCompressed && cacheOnRead) {
+        if (cacheOnly && cacheCompressed && cacheOnRead && shouldCacheFileBlock) {
```

Comment:
Still in the same line of my previous comments: why do we need these two separate booleans (cacheOnRead && shouldCacheFileBlock) here? We should merge the data tiering logic inside shouldCacheBlockOnRead and deal with a single boolean variable.

Reply:
ack. I have added an overloaded function to be used in this case and kept the original function to be used at other places. Please take a look.

Thanks,
Janardhan